service_client/service_client.go

```diff
@@ -0,0 +1,169 @@
+	return NewServiceClientWithOpts(serviceURL, ServiceClientOpts{})
+}
+
+func NewServiceClientWithOpts(serviceURL string, opts ServiceClientOpts) *serviceClient {
+	u, err := url.Parse(serviceURL)
+	if err != nil {
+		panic(err)
+	}
+	httpClient := &http.Client{Transport: AggressiveTransport}
+	client := httpx.NewServiceClient(serviceURL, httpClient)
+	signer := httpsignatures.DefaultSha256Signer
+	if opts.Scrubber == nil {
+		opts.Scrubber = &NoopScrubber{}
+	}
+
+	return &serviceClient{
+		serviceURL:       u,
+		client:           client,
+		signer:           signer,
+		keyId:            opts.SigningKeyId,
+		key:              opts.SigningKey,
+		forwardedHeaders: opts.IncludeForwardedHeaders,
+		logScrubber:      opts.Scrubber,
+	}
+}
+
+func (c *serviceClient) Do(ctx context.Context, method, path string, jsonData interface{}, targetObject interface{}) error {
+	return c.do(ctx, method, path, "", jsonData, targetObject)
+}
+
+func (c *serviceClient) DoWithBearerAuth(ctx context.Context, method, path, token string, jsonData interface{}, targetObject interface{}) error {
+	return c.do(ctx, method, path, token, jsonData, targetObject)
+}
+
+func (c *serviceClient) do(ctx context.Context, method, path, token string, jsonData interface{}, targetObject interface{}) error {
+	requestPath := httpx.ParseURL(c.serviceURL, path)
+	req, err := httpx.NewJSONRequest(method, requestPath, jsonData)
+
+	if err != nil {
+		return err
+	}
+
+	if token == "" {
+		c.setBasicAuth(req)
+	} else {
+		c.setBearerAuth(req, token)
+	}
+	if c.key != "" {
+		c.setHttpSignature(req)
+	}
+
+	for _, header := range c.forwardedHeaders {
+		req.Header.Add(header, httpx.Header(ctx, header))
+	}
+	traceCloser := c.trace(ctx, req)
+
+	resp, err := c.client.Do(ctx, req)
+	traceCloser()
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+
+	if err != nil {
+		return err
+	}
+	if err = c.checkResponse(resp); err != nil {
+		return err
+	}
+	if targetObject == nil {
+		_, err := io.Copy(ioutil.Discard, resp.Body)
+		return err
+	}
+
+	return json.NewDecoder(resp.Body).Decode(targetObject)
+}
+
+func (c *serviceClient) setBasicAuth(req *http.Request) {
+	if c.serviceURL.User == nil {
+		return
+	}
+
+	password, _ := c.serviceURL.User.Password()
+	req.SetBasicAuth(c.serviceURL.User.Username(), password)
+}
+
+func (c *serviceClient) trace(ctx context.Context, req *http.Request) func() {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "client.request")
+	opentracing.GlobalTracer().Inject(
+		span.Context(),
+		opentracing.HTTPHeaders,
+		opentracing.HTTPHeadersCarrier(req.Header))
+	span.SetTag("uri", c.logScrubber.Scrub(req.URL.String()))
+	return func() {
+		span.Finish()
+	}
+}
+
+func (c *serviceClient) setHttpSignature(req *http.Request) error {
+	err := c.signer.SignRequest(c.keyId, c.key, req)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *serviceClient) setBearerAuth(req *http.Request, token string) {
```

Comment:
Bearer auth vs Basic auth.

Basic auth set at the client level is useful for static credentials. Bearer auth allows supplying a token at call time, so this client can be used with 3rd party services.